contracts/src/Nodes.sol

```diff
@@ -1,91 +1,162 @@
 // SPDX-License-Identifier: MIT
 pragma solidity ^0.8.13;
 
+import "@openzeppelin/contracts/token/ERC721/ERC721.sol";
 import "@openzeppelin/contracts/access/Ownable.sol";
 
-contract Nodes is Ownable {
-    constructor() Ownable(msg.sender) {}
+/**
+A NFT contract for XMTP Node Operators.
 
+The deployer of this contract is responsible for minting NFTs and assinging them to node operators.
+
+All nodes on the network periodically check this contract to determine which nodes they should connect to.
+ */
+contract Nodes is ERC721, Ownable {
+    constructor() ERC721("XMTP Node Operator", "XMTP") Ownable(msg.sender) {}
+
+    // uint16 counter so that we cannot create more than 65k IDs
+    // The ERC721 standard expects the tokenID to be uint256 for standard methods unfortunately
+    uint16 private _nodeIdCounter;
+
+    // A node, as stored in the internal mapping
     struct Node {
+        bytes signingKeyPub;
         string httpAddress;
-        uint256 originatorId;
         bool isHealthy;
-        // Maybe we want a TLS cert separate from the public key for MTLS authenticated connections?
     }
 
-    event NodeUpdate(
-        bytes publicKey,
-        string httpAddress,
-        uint256 originatorId,
-        bool isHealthy
-    );
+    struct NodeWithId {
+        uint16 nodeId;
+        Node node;
+    }
 
-    // List of public keys
-    bytes[] publicKeys;
+    event NodeUpdated(uint256 nodeId, Node node);
 
-    // Mapping of publicKey to node
-    mapping(bytes => Node) public nodes;
+    // Mapping of token ID to Node
+    mapping(uint256 => Node) private _nodes;
 
     /**
-    Add a node to the network
+    Mint a new node NFT and store the metadata in the smart contract
      */
     function addNode(
-        bytes calldata publicKey,
+        address to,
+        bytes calldata signingKeyPub,
+        string calldata httpAddress
+    ) public onlyOwner returns (uint16) {
+        uint16 nodeId = _nodeIdCounter;
```

Comment:
We should not allow a node to have node ID 0, because an originator SID of 0 means that the originator is the blockchain. Wondering if we should pre-increment _nodeIdCounter rather than post-incrementing it?

Reply:
I have a comment somewhere about this. I kinda think we should start at 1000 just to be safe. Would let us partition the smart contracts later if we wanted to.